x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskTests.java

```diff
@@ -5,205 +5,613 @@
-    }
-
-    public void testNotAllExpectedOpsReturned() throws Exception {
-        task = createShardFollowTask(1024, 1, 1, 10000, 1024, -1);
-        task.start(-1);
-        randomlyTruncateRequests.set(true);
-        assertBusy(() -> {
-            assertThat(task.getStatus().getFollowerGlobalCheckpoint(), equalTo(10000L));
-        });
-        assertThat(truncatedRequests.get(), greaterThan(0));
-    }
-
-    ShardFollowNodeTask createShardFollowTask(int maxBatchOperationCount, int maxConcurrentReadBatches, int maxConcurrentWriteBathces,
-                                              int globalCheckpoint, int bufferWriteLimit, long followGlobalCheckpoint) {
-        leaderGlobalCheckPoint = new AtomicLong(globalCheckpoint);
-        imdVersion = new AtomicLong(1L);
-        mappingUpdateCounter = new AtomicInteger(0);
-        randomlyTruncateRequests = new AtomicBoolean(false);
-        truncatedRequests = new AtomicInteger();
-        randomlyFailWithRetryableError = new AtomicBoolean(false);
-        failedRequests = new AtomicInteger(0);
+    private Exception fatalError;
+    private List<long[]> shardChangesRequests;
+    private List<List<Translog.Operation>> bulkShardOperationRequests;
+    private BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> task.run();
+
+    private Queue<Exception> readFailures;
+    private Queue<Exception> writeFailures;
+    private Queue<Exception> mappingUpdateFailures;
+    private Queue<Long> imdVersions;
+    private Queue<Long> followerGlobalCheckpoints;
+
+    public void testCoordinateReads() {
+        ShardFollowNodeTask task = createShardFollowTask(8, 8, 8, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(8));
+        assertThat(shardChangesRequests, contains(new long[][]{
+            {0L, 8L}, {8L, 8L}, {16L, 8L}, {24L, 8L}, {32L, 8L}, {40L, 8L}, {48L, 8L}, {56L, 8L}}
+        ));
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(8));
+        assertThat(status.getLastRequestedSeqno(), equalTo(63L));
+    }
+
+    public void testWriteBuffer() {
+        // Need to set concurrentWrites to 0, other the write buffer gets flushed immediately:
+        ShardFollowNodeTask task = createShardFollowTask(64, 1, 0, 32, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(1));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(64L));
+
+        shardChangesRequests.clear();
+        // Also invokes the coordinatesReads() method:
+        task.innerHandleReadResponse(0L, 63L, generateShardChangesResponse(0, 63, 0L, 128L));
+        assertThat(shardChangesRequests.size(), equalTo(0)); // no more reads, because write buffer is full
```

Comment:
I am wondering if we should add `buffer` (size or operations) into the Status object? We can do it in a follow up if you are okay.

Reply:
I think we should add this too, but in a follow up. I think when Jason is going to work on the ccr stats then this is a stat that is very likely to be added.